boa/src/vm/mod.rs

```diff
@@ -736,18 +882,52 @@ impl Context {
                 self.vm.push(result);
             }
             Opcode::Return => {
-                for _ in 0..self.vm.frame().pop_env_on_return {
-                    self.pop_environment();
+                if let Some(finally_address) = self.vm.frame().catch.last().and_then(|c| c.finally)
+                {
+                    let frame = self.vm.frame_mut();
+                    frame.pc = finally_address as usize;
+                    frame.finally_return = FinallyReturn::Ok;
+                    frame.catch.pop();
+                } else {
+                    for _ in 0..self.vm.frame().pop_env_on_return {
+                        self.pop_environment();
+                    }
+                    self.vm.frame_mut().pop_env_on_return = 0;
+                    let _ = self.vm.pop_frame();
+                    return Ok(true);
                 }
-                self.vm.frame_mut().pop_env_on_return = 0;
-                let _ = self.vm.pop_frame();
-                return Ok(true);
             }
             Opcode::PushDeclarativeEnvironment => {
                 let env = self.get_current_environment();
                 self.push_environment(DeclarativeEnvironmentRecord::new(Some(env)));
                 self.vm.frame_mut().pop_env_on_return += 1;
             }
+            Opcode::PushFunctionEnvironment => {
+                let is_constructor = self.vm.frame().code.constructor;
+                let is_lexical = self.vm.frame().code.this_mode.is_lexical();
+                let current_env = self.get_current_environment();
+                let this = &self.vm.frame().this;
+
+                let new_env = FunctionEnvironmentRecord::new(
+                    this.clone().as_object().unwrap().clone(), //TODO: is this ok? this_function object on stack mb?
```

Comment:
Is it not possible for `this` to not be an object? For example, with `Function.prototype.call`, the user can choose the `this` with which the function is invoked, wouldn't that result in possible panics?

Reply:
Good catch. I looked at the spec a bit. `this` should always be an object as you can see in step 6 of [`OrdinaryCallBindThis`](https://tc39.es/ecma262/#sec-ordinarycallbindthis), that is called for call and construct. I added that conversion.